File: python-sdks/examples/openai-sdk/attributes.py
"""Attributes — Attach customer info and metadata to traces."""

import os
from dotenv import load_dotenv

load_dotenv(override=True)

from openai import OpenAI
from respan import Respan, workflow, propagate_attributes
from respan_instrumentation_openai import OpenAIInstrumentor

respan = Respan(
    instrumentations=[OpenAIInstrumentor()],
    metadata={"service": "chat-api", "version": "1.0.0"},
)

client = OpenAI(
    api_key=os.getenv("RESPAN_API_KEY"),
    base_url=os.getenv("RESPAN_BASE_URL", "https://api.respan.ai/api"),
)


@workflow(name="handle_request")
def handle_request(user_id: str, question: str):
    with propagate_attributes(
        customer_identifier=user_id,
        thread_identifier="conv_001",
        metadata={"plan": "pro"},  # merged with default metadata
    ):
        response = client.chat.completions.create(
            model="gpt-4.1-nano",
            messages=[{"role": "user", "content": question}],

        )
        print(f"[{user_id}] {response.choices[0].message.content}")


handle_request("user_alice", "What is an API gateway?")
handle_request("user_bob", "Explain rate limiting.")
respan.flush()

File: javascript-sdks/examples/openai-sdk/attributes.ts
/** Attributes — Attach customer info and metadata to traces. */

import "dotenv/config";
import OpenAI from "openai";
import { Respan } from "@respan/respan";
import { OpenAIInstrumentor } from "@respan/instrumentation-openai";

const respan = new Respan({
  apiKey: process.env.RESPAN_API_KEY,
  baseURL: process.env.RESPAN_BASE_URL,
  instrumentations: [new OpenAIInstrumentor()],
});
await respan.initialize();

const client = new OpenAI({
  apiKey: process.env.RESPAN_API_KEY,
  baseURL: process.env.RESPAN_BASE_URL,
});

async function handleRequest(userId: string, question: string) {
  await respan.propagateAttributes(
    {
      customer_identifier: userId,
      thread_identifier: "conv_001",
      metadata: { plan: "pro" },
    },
    async () => {
      return respan.withWorkflow({ name: "handle_request" }, async () => {
        const response = await client.chat.completions.create({
          model: "gpt-4.1-nano",
          messages: [{ role: "user", content: question }],
        });
        console.log(`[${userId}] ${response.choices[0].message.content}`);
        return response;
      });
    }
  );
}

await handleRequest("user_alice", "What is an API gateway?");
await handleRequest("user_bob", "Explain rate limiting.");
